File: src/java/EntityFX/src/EntityFX/Core/Whetstone/ParallelWhetstoneBenchmark.java
package EntityFX.Core.Whetstone;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Arrays;
import java.util.concurrent.ExecutionException;

import EntityFX.Core.Writer;
import EntityFX.Core.Generic.BenchResult;
import EntityFX.Core.Generic.BenchmarkBase;

public class ParallelWhetstoneBenchmark extends BenchmarkBase<BenchResult[]> {

    public ParallelWhetstoneBenchmark(Writer writer, boolean printToConsole) throws FileNotFoundException {
        super(writer, printToConsole);
        isParallel = true;
    }

    @Override
    public BenchResult[] benchImplementation() throws InterruptedException, ExecutionException {
        return benchInParallel(() -> {
            try {
                return new Whetstone(false);
            } catch (FileNotFoundException e1) {
                e1.printStackTrace();
            }
            return null;
        }, a -> {
            try {
                return a.bench(false);
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }, (a, r) -> {
            r.Points = a.getMWIPS() * Ratio;
            r.Result = a.getMWIPS();
            r.Output = a.getOutput();
        });
    }

    @Override
    public BenchResult populateResult(BenchResult benchResult, BenchResult[] whetstonBenchResults) {
        var result = buildParallelResult(benchResult, whetstonBenchResults);
        result.Result = Arrays.stream(whetstonBenchResults).mapToDouble(r -> r.Result).sum();
        result.Units = "MWIPS";
        result.Output = String.join("", Arrays.toString(Arrays.stream(whetstonBenchResults).map(r -> r.Output).toArray()));
        return result;
    }

    @Override
    public void warmup(Double aspect) {
    }
}

File: src/python/entityfx/parallel_whetstone_benchmark.py
from entityfx.whetstone import Whetstone
from entityfx.benchmark_base import BenchmarkBase
from entityfx.writer import Writer

class ParallelWhetstoneBenchmark(BenchmarkBase):

    def __init__(self, writer: Writer, print_to_console : bool=True) -> None:
        super().__init__(writer, print_to_console)
        self.is_parallel = True
        self.ratio = 1
    
    def benchImplementation(self) -> list:
        return self.bench_in_parallel(lambda : Whetstone(False), lambda a: a.bench(), lambda a, r : self.setBenchResultFunc(a, r))

    def warmup(self, aspect : float=.05) -> None:
        pass

    def populateResult(self, bench_result, results : list):
        result = self._buildParallelResult(bench_result, results) 
        result["Result"] = sum(map(lambda x : x["Result"], results))
        result["Units"] = "MWIPS"
        result["Output"] = "".join(map(lambda x : x["Output"], results))
        return result

    def setBenchResultFunc(self, a, r):
        r["Points"] = a["MWIPS"] * self.ratio
        r["Result"] = a["MWIPS"]
        r["Output"] = a["Output"]